### PR TITLE
Add GPU version of ToPrettyString [databricks]

### LINKED ISSUE #9065: [BUG][Spark 3.5.0] Reinstate cast map/array to string tests when 3.5.0-rc2 is available
Describe the bug
`test_cast_map_to_string` and `test_cast_array_to_string` fail due to a bug in Spark 3.5.0-rc1, so the test is xfailed in https://github.com/NVIDIA/spark-rapids/pull/8969 and we need to reinstate the test once 3.5.0-rc2 is available

Steps/Code to reproduce bug

Expected behavior

Environment details (please complete the following information)

Additional context

### LINKED ISSUE #9171: [FEA] Add GPU version of ToPrettyString
**Is your feature request related to a problem? Please describe.**
Spark 3.5 adds a new Expression `ToPrettyString` that acts similar to cast to string, except it prints `NULL` for NULL values, a HEX string for binary values, and a slightly different format for decimal values. We don't currently support this on the GPU which causes show to fall back to the CPU.

https://github.com/apache/spark/blob/40943c2748fdd28d970d017cb8ee86c294ee62df/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ToPrettyString.scala#L26-L31

**Describe the solution you'd like**
Have an equivalent GPU version so we can replace the ToPrettString expression. 

**Describe alternatives you've considered**
No alternatives were tried

### LINKED ISSUE #9220: [FEA] Add GPU support for converting binary data to a hex string in REPL
**Is your feature request related to a problem? Please describe.**
In Spark 3.5+ REPL displays binary data as hex strings when a projection contains binary data. We currently are falling back to the CPU as `GpuToPrettyString` doesn't support binary to hex string conversion

**Describe the solution you'd like**
Once https://github.com/rapidsai/cudf/issues/14081 is done we can add this functionality so SparkShims doesn't make `GpuToPrettyString` fall back to CPU.

### PR DESCRIPTION
**Pull request by razajafri**
This PR adds a `GpuToPrettyString` so we don't fall back to the CPU when `show()` is called in REPL. 

The PR was manually tested and unit tests were added 

fixes #9065 
fixes #9171 
fixes #9220

### CODE REVIEW

File: sql-plugin/src/main/spark350/scala/com/nvidia/spark/rapids/shims/SparkShims.scala [hunk 1]
```diff
@@ -21,13 +21,29 @@ package com.nvidia.spark.rapids.shims
...
 object SparkShimImpl extends Spark340PlusShims {
 
   override def getExprs: Map[Class[_ <: Expression], ExprRule[_ <: Expression]] = {
     val shimExprs: Map[Class[_ <: Expression], ExprRule[_ <: Expression]] = Seq(
+      GpuOverrides.expr[ToPrettyString]("An internal expressions which is used to " +
+        "generate pretty string for all kinds of values",
+        ExprChecks.unaryProject(TypeSig.STRING, TypeSig.STRING, TypeSig.all, TypeSig.all),
+        (a, conf, p, r) => new UnaryExprMeta[ToPrettyString](a, conf, p, r) {
+          override def tagExprForGpu(): Unit = {
+            if (a.asInstanceOf[ToPrettyString].child.dataType == BinaryType) {
```
Comment (revans2) [inline sql-plugin/src/main/spark350/scala/com/nvidia/spark/rapids/shims/SparkShims.scala:37]: Why do we say that we support all types as input in the TypeSig when we don't support Binary?  What about an ARRAY[Binary] Do we support that? This is not checking for that when the TypeSig would. What about user defined types?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala [hunk 2]
```diff
@@ -1482,88 +1258,4 @@ object GpuCast {
       _.castTo(GpuColumnVector.getNonNestedRapidsType(toType))
     }
   }
-}
-
-/**
- * Casts using the GPU
- */
-case class GpuCast(
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala:1490]: Why move this class? It make checking for changes to the code a lot more difficult to do.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/ToStringBase.scala [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+import scala.collection.mutable.ArrayBuffer
+
+import ai.rapids.cudf.{BinaryOp, ColumnVector, ColumnView, DType, RegexProgram, Scalar}
+import com.nvidia.spark.rapids.Arm.{closeOnExcept, withResource}
+import com.nvidia.spark.rapids.RapidsPluginImplicits._
+import com.nvidia.spark.rapids.shims.GpuCastShims
+
+import org.apache.spark.sql.types._
+
+trait ToStringBase {
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/ToStringBase.scala:30]: Could you add docs to explain what the point to ToStringBase and how it relates to CastOperation?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala [hunk 4]
```diff
@@ -163,7 +162,90 @@ final class CastExprMeta[INPUT <: UnaryExpression with TimeZoneAwareExpression w
...
+    case _: ArrayType | _: MapType | _: StructType => child.sql
+    case _ => s"CAST(${child.sql} AS ${dataType.sql})"
+  }
+
+  override def doColumnar(input: GpuColumnVector): ColumnVector =
+    CastOperation(input.getBase, input.dataType(), dataType, ansiMode,
+      legacyCastToString, stringToDateAnsiModeEnabled)
+}
+
+object CastOperation {
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala:248]: If the point of ToStringBase is to provide a clean way to cast something to a String, then why does the CastOperation object not inherent the ToStringBase trait? And if CastOperation is really static, and Specific to how GpuCast works, then why is it not still called GpuCast?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala [hunk 5]
```diff
@@ -1366,33 +1169,6 @@ object GpuCast {
     }
   }
 
-  def fixDecimalBounds(input: ColumnView,
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala:1369]: Why did you move these functions? It makes it much harder to review the code when things move around.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala [hunk 6]
```diff
@@ -187,27 +269,88 @@ object GpuCast {
...
+  }
+
+  def apply(
       input: ColumnView,
       fromDataType: DataType,
       toDataType: DataType,
       ansiMode: Boolean,
       legacyCastToString: Boolean,
       stringToDateAnsiModeEnabled: Boolean): ColumnVector = {
+    new CastOperation(input, fromDataType, toDataType, ansiMode,
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala:310]: I don't see any value in creating this. An object can implement a trait. We don't need a CastOperation class. It is just adding overhead.

**Review by revans2 — COMMENTED** [review]
I did a quick pass through much of the code and it feels like there is a lot of refactoring that I have a hard time seeing any value with. I understand ToStringBase so you can override how some functions are implemented, but everything else feels unnecessary.

File: sql-plugin/src/main/spark350/scala/com/nvidia/spark/rapids/shims/GpuToPrettyString.scala [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+      GpuColumnVector.from(stringCol, StringType)
+    }
+  }
+
+  override def columnarEvalAny(batch: ColumnarBatch): Any = {
+    val c = child.columnarEvalAny(batch)
+    if (c.isInstanceOf[GpuColumnVector]) {
+      processColumnVector(c.asInstanceOf[GpuColumnVector])
+    } else {
+      withResource(c.asInstanceOf[Scalar]) { scalar =>
```
Comment (revans2) [inline sql-plugin/src/main/spark350/scala/com/nvidia/spark/rapids/shims/GpuToPrettyString.scala:79]: This does not look correct at all to me. Why would Spark create their own `ToStringBase` if they could just get the scala/java object and call `toString` on it? Please don't implement/call `columnarEvalAny` unless you really can handle scalar values properly.  Just implement and call `columnarEval`

File: tests/src/test/spark350/scala/com/nvidia/spark/rapids/ToPrettyStringSuite.scala [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+      StructField("c", DataTypes.IntegerType)))
+    val v = Array(1, 2, 3)
+    val cpuResult = executeOnCpuAndReturn(dataType, InternalRow(v: _*))
+    val child = GpuBoundReference(0, dataType, true)(NamedExpression.newExprId, "struct_arg")
+    val gpuToPrettyStr = GpuToPrettyString(child, Some("UTC"))
+    val l =
+      List[HostColumnVector.DataType](
+        new HostColumnVector.BasicType(false, DType.INT32),
+        new HostColumnVector.BasicType(false, DType.INT32),
+        new HostColumnVector.BasicType(false, DType.INT32),
```
Comment (andygrove) [inline tests/src/test/spark350/scala/com/nvidia/spark/rapids/ToPrettyStringSuite.scala:125]: Trailing comma causes scalastyle to fail

```suggestion
        new HostColumnVector.BasicType(false, DType.INT32)
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala [hunk 9]
```diff
@@ -263,8 +273,8 @@ object GpuCast {
         withResource(input.min()) { minInput =>
           withResource(input.max()) { maxInput =>
             if (minInput.isValid && minInput.getBigDecimal().compareTo(min) == -1 ||
-                maxInput.isValid && maxInput.getBigDecimal().compareTo(max) == 1) {
-              throw new ArithmeticException(GpuCast.OVERFLOW_MESSAGE)
+              maxInput.isValid && maxInput.getBigDecimal().compareTo(max) == 1) {
```
Comment (andygrove) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala:276]: The original indentation was clearer to me, to indicate continuation of the expression from the previous line.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/ToStringBase.scala [hunk 10]
```diff
@@ -0,0 +1,388 @@
...
+    val (leftStr, rightStr) = ("[", "]")
+    val emptyStr = ""
+    val nullStr = nullString
+    val numRows = input.getRowCount.toInt
+
+    withResource(
+      Seq(leftStr, rightStr, emptyStr, nullStr).safeMap(Scalar.fromString)
+    ) { case Seq(left, right, empty, nullRep) =>
+      val strChildContainsNull = withResource(input.getChildColumnView(0)) { child =>
+        GpuCast.doCast(child, elementType, StringType, ansiMode,
```
[the remark this answers is not in the document]
Comment (razajafri) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/ToStringBase.scala:224]: It doesn't matter but for clarity, I will make the change

File: sql-plugin/src/main/spark350/scala/com/nvidia/spark/rapids/shims/GpuToPrettyString.scala [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+  override protected def useDecimalPlainString: Boolean = true
+
+  override protected def useHexFormatForBinary: Boolean = true
+
+  private def processColumnVector(cv: GpuColumnVector) = {
+    val stringCol = withResource(cv) { originalCol =>
+      castToString(originalCol.getBase, originalCol.dataType(), false, false, false)
+    }
+
+    if (stringCol.hasNulls()) {
```
[the remark this answers is not in the document]
Comment (razajafri) [reply sql-plugin/src/main/spark350/scala/com/nvidia/spark/rapids/shims/GpuToPrettyString.scala:61]: This check isn't necessary as the nulls will be converted to string as you correctly have identified

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala [hunk 12]
```diff
@@ -187,14 +185,28 @@ object GpuCast {
...
+
+  override protected def rightBracket: String =
+    if (SQLConf.get.getConf(SQLConf.LEGACY_COMPLEX_TYPES_TO_STRING)) "]" else "}"
+
+  override protected def nullString: String =
+    if (SQLConf.get.getConf(SQLConf.LEGACY_COMPLEX_TYPES_TO_STRING)) "" else "null"
+
+  // In ANSI mode, Spark always uses plain string representation on casting Decimal values
+  // as strings. Otherwise, the casting may use scientific notation if an exponent is needed.
+  override protected def useDecimalPlainString: Boolean = SQLConf.get.getConf(SQLConf.ANSI_ENABLED)
```
[the remark this answers is not in the document]
Comment (razajafri) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala:199]: These are functions and will return the latest value

**Comment by razajafri** [comment]
@revans2 and @andygrove thank you for reviewing. 

I have addressed all of your comments I think. PTAL

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala [hunk 13]
```diff
@@ -187,14 +202,29 @@ object GpuCast {
 
   val INVALID_NUMBER_MSG: String = "At least one value is either null or is an invalid number"
 
+  private def ansiMode = SQLConf.get.getConf(SQLConf.ANSI_ENABLED)
+
+  private def legacyCastString = SQLConf.get.getConf(SQLConf.LEGACY_COMPLEX_TYPES_TO_STRING)
```
Comment (andygrove) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala:207]: These variables will only get initialized once, using the config at that time. The config could change between different queries being executed. The `doCastr` method already has `ansiMode` and `legacyCastToString` passed in, so I'm not sure why we need them here as well.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala [hunk 14]
```diff
@@ -1506,6 +1506,21 @@ class CastChecks extends ExprChecks {
...
+/** 
+ * This class is just restricting the 'to' dataType to a StringType in the CastChecks class
+ */ 
+class ToPrettyStringChecks extends CastChecks {
+
+  override protected def tagBase(meta: RapidsMeta[_, _, _], willNotWork: String => Unit): Unit = {
+    val cast = meta.wrapped.asInstanceOf[UnaryExpression]
+    val from = cast.child.dataType
+    val to = StringType
+    if (!gpuCanCast(from, to)) {
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala:1518]: I don't think this is going to show up in the generated docs, even if we tried to build them with 3.5.0. We should at least have a follow on issue to figure that out.

**Review by revans2 — COMMENTED** [review]
The only non-nit I have left is what Andy showed about ANSI mode for cast.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala (hunk 13 above)
Comment (razajafri) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala:207]: There are instance variables that are dependent on the conf, they are outside of the doCast function. 

These `variables` are functions and will be updated every time an instance is made and thus will pick up the new conf. Am I missing something?

Comment (andygrove) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala:207]: I did miss that these are `def` not `val`, but it seems odd to have them defined both as variables at the object level and also have them passed in to the `doCast` method.

Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala:207]: If I do an `ansi_cast` instead of a `cast`, Does that mean that `useDecimalPlainString` will be wrong? Or what about if I change  `spark.sql.ansi.enabled` part way through a session?  It is perfectly valid to change it. Would that impact the result of `useDecimalPlainString` and possible return the wrong result for a decimal value?  `legacyCastToString is similar. I can change it at runtime.

Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala:207]: Sorry I missed the part about them being functions. The big issue is then `ansi_cast`, where I can have a Cast expression that is marked as ANSI, but the config is not set for ANSI.

Comment (razajafri) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala:207]: I didn't realize there was a way to invoke ANSI without setting the conf. If that is the case we definitely have a problem. Can you please give me an example? The only way I could come up with is to write a hive table and insert decimal values into a string column.

Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala:207]: Sorry `ansi_cast` is not a real operator. The point is that the API provides one set of expectations, but does not live up to them.  That for me is a bug.  We have our own code that violates this constraint.

https://github.com/NVIDIA/spark-rapids/blob/01fce6c909cae55bd1b8146cc63e9ec055316ad7/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala#L384

What is to stop spark from doing it in the future?  It makes the code brittle.

Comment (razajafri) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala:207]: I agree, I am not too happy with this situation either now that you mention it. I designed it to work with the `CastOperation` solution but now I am just trying to contort this into position. I need to take a step back and think this through

File: tests/src/test/spark350/scala/com/nvidia/spark/rapids/ToPrettyStringSuite.scala [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+import com.nvidia.spark.rapids.shims.GpuToPrettyString
+
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{Literal, NamedExpression, ToPrettyString}
+import org.apache.spark.sql.catalyst.util.{ArrayBasedMapData, ArrayData}
+import org.apache.spark.sql.types.{ArrayType, DataType, DataTypes, MapType, StructField, StructType}
+import org.apache.spark.sql.vectorized.ColumnarBatch
+import org.apache.spark.unsafe.types.UTF8String
+
+class ToPrettyStringSuite extends GpuUnitTests {
```
Comment (andygrove) [inline tests/src/test/spark350/scala/com/nvidia/spark/rapids/ToPrettyStringSuite.scala:34]: Could you add tests for other types, such as decimal, float/double, int types, string, covering null values. I noticed that we have different behavior for null decimals through manual testing, where Spark shows `NULL` and on GPU we show an empty string instead.

## CPU

```
scala> spark.sql("select ss_ext_sales_price from store_sales").show
+------------------+
|ss_ext_sales_price|
+------------------+
|            817.20|
|            117.04|
|           1189.20|
|            239.12|
|           1260.87|
|            185.13|
|            462.95|
|            122.94|
|             31.80|
|            111.29|
|              NULL|
|            131.25|
|            858.00|
|           1247.38|
|           3011.16|
|           2599.80|
|           2565.16|
|            343.06|
|           4918.32|
|             29.56|
+------------------+
only showing top 20 rows
```

## GPU 

```
scala> spark.sql("select ss_ext_sales_price from store_sales").show
23/09/21 16:46:04 WARN GpuOverrides: 
!Exec <CollectLimitExec> cannot run on GPU because the Exec CollectLimitExec has been disabled, and is disabled by default because Collect Limit replacement can be slower on the GPU, if huge number of rows in a batch it could help by limiting the number of rows transferred from GPU to CPU. Set spark.rapids.sql.exec.CollectLimitExec to true if you wish to enable it
  @Partitioning <SinglePartition$> could run on GPU
  *Exec <ProjectExec> will run on GPU
    *Expression <Alias> toprettystring(ss_ext_sales_price#15, Some(UTC)) AS toprettystring(ss_ext_sales_price)#251 will run on GPU
      *Expression <ToPrettyString> toprettystring(ss_ext_sales_price#15, Some(UTC)) will run on GPU
    *Exec <FileSourceScanExec> will run on GPU

+------------------+
|ss_ext_sales_price|
+------------------+
|            817.20|
|            117.04|
|           1189.20|
|            239.12|
|           1260.87|
|            185.13|
|            462.95|
|            122.94|
|             31.80|
|            111.29|
|                  |
|            131.25|
|            858.00|
|           1247.38|
|           3011.16|
|           2599.80|
|           2565.16|
|            343.06|
|           4918.32|
|             29.56|
+------------------+
only showing top 20 rows
```

Comment (andygrove) [reply tests/src/test/spark350/scala/com/nvidia/spark/rapids/ToPrettyStringSuite.scala:34]: I also found some formatting differences with float and decimal that we should at least document if we cannot easily produce the same output, and file a follow-on issue.

```
scala> spark.conf.set("spark.rapids.sql.enabled", false)
scala> df2.show
+--------------------+------------+--------------------+
|                   a|           b|                   c|
+--------------------+------------+--------------------+
| 9223372036854775807| 9.223372E18|9.223372036854776E18|
|-9223372036854775808|-9.223372E18|-9.22337203685477...|
+--------------------+------------+--------------------+

scala> spark.conf.set("spark.rapids.sql.enabled", true)
scala> df2.show

+--------------------+---------------+---------------+
|                   a|              b|              c|
+--------------------+---------------+---------------+
| 9223372036854775807| 9.223372037E18| 9.223372037E18|
|-9223372036854775808|-9.223372037E18|-9.223372037E18|
+--------------------+---------------+---------------+
```

Comment (razajafri) [reply tests/src/test/spark350/scala/com/nvidia/spark/rapids/ToPrettyStringSuite.scala:34]: This is an existing issue. I just checked with Spark-3.4 using spark-rapids version 23.08 

```
val df = Seq(9223372036854775807L, -9223372036854775808L).toDF("a").repartition(2)
val df2 = df.withColumn("b", expr("cast(a as float)")).withColumn("c", expr("cast(a as double)"))

spark.conf.set("spark.rapids.sql.enabled", false)
df2.show

+--------------------+------------+--------------------+
|                   a|           b|                   c|
+--------------------+------------+--------------------+
| 9223372036854775807| 9.223372E18|9.223372036854776E18|
|-9223372036854775808|-9.223372E18|-9.22337203685477...|
+--------------------+------------+--------------------+

spark.conf.set("spark.rapids.sql.enabled", true)
df2.show

+--------------------+---------------+---------------+
|                   a|              b|              c|
+--------------------+---------------+---------------+
| 9223372036854775807| 9.223372037E18| 9.223372037E18|
|-9223372036854775808|-9.223372037E18|-9.223372037E18|
+--------------------+---------------+---------------+
```

Comment (razajafri) [reply tests/src/test/spark350/scala/com/nvidia/spark/rapids/ToPrettyStringSuite.scala:34]: This is also giving me a hard time to write the unit tests for float types

Comment (razajafri) [reply tests/src/test/spark350/scala/com/nvidia/spark/rapids/ToPrettyStringSuite.scala:34]: This is already being tracked in https://github.com/NVIDIA/spark-rapids/issues/4204. I have added this example to the ticket.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala (hunk 14 above)
Comment (razajafri) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala:1518]: Wondering if there is a valid need to print `ToPrettyString` as part of the `supported_ops`. It is a sub set of what `GpuCast` can do.

**Comment by razajafri** [comment]
@andygrove @revans2 PTAL.

**Comment by razajafri** [comment]
build

File: tests/src/test/spark350/scala/com/nvidia/spark/rapids/ToPrettyStringSuite.scala [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+    // This test is expected to fail until https://github.com/NVIDIA/spark-rapids/issues/4204
+    // is resolved
+    assertThrows[TestFailedException](testDataType(DataTypes.DoubleType))
+  }
+
+  test("test ints") {
+    testDataType(DataTypes.IntegerType)
+  }
+
+  test("test longs") {
```
Comment (andygrove) [inline tests/src/test/spark350/scala/com/nvidia/spark/rapids/ToPrettyStringSuite.scala:191]: Could you add tests for the remaining primitive types: char, byte, short, boolean

Comment (razajafri) [reply tests/src/test/spark350/scala/com/nvidia/spark/rapids/ToPrettyStringSuite.scala]: I have added tests can you PTAL

File: tests/src/test/scala/com/nvidia/spark/rapids/GpuBatchUtilsSuite.scala [hunk 17]
```diff
@@ -25,6 +25,7 @@ import org.scalatest.funsuite.AnyFunSuite
 import org.apache.spark.sql.Row
```
Comment (razajafri) [inline tests/src/test/scala/com/nvidia/spark/rapids/GpuBatchUtilsSuite.scala:25]: I had to make changes to this file because of the way InternalRow accesses Arrays, Maps and Structs when using `BoundReference`

**Review by andygrove — APPROVED** [review]

**Review by revans2 — APPROVED** [review]

**Comment by andygrove** [comment]
@mimaomao @kevnzhao fyi, this PR is ready to go, but we are just waiting on https://github.com/rapidsai/cudf/pull/14205 to be merged before we merge this one.

**Comment by razajafri** [comment]
build
